src/main/java/com/google/devtools/build/lib/skyframe/TargetCycleReporter.java

```diff
@@ -13,49 +13,73 @@
 // limitations under the License.
 package com.google.devtools.build.lib.skyframe;
 
+import com.google.common.base.Preconditions;
 import com.google.common.base.Predicate;
 import com.google.common.base.Predicates;
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.Iterables;
 import com.google.common.collect.Lists;
+import com.google.devtools.build.lib.actions.ActionLookupKey;
 import com.google.devtools.build.lib.cmdline.Label;
 import com.google.devtools.build.lib.events.ExtendedEventHandler;
 import com.google.devtools.build.lib.packages.PackageGroup;
 import com.google.devtools.build.lib.packages.Target;
 import com.google.devtools.build.lib.pkgcache.PackageProvider;
+import com.google.devtools.build.lib.skyframe.AspectValueKey.AspectKey;
 import com.google.devtools.build.skyframe.CycleInfo;
 import com.google.devtools.build.skyframe.SkyKey;
 import java.util.List;
 
 /**
- * Reports cycles between {@link TransitiveTargetValue}s. These indicates cycles between targets
- * (e.g. '//a:foo' depends on '//b:bar' and '//b:bar' depends on '//a:foo').
+ * Reports cycles between targets. These may be in the form of {@link ConfiguredTargetValue}s or
+ * {@link TransitiveTargetValue}s.
  */
-class TransitiveTargetCycleReporter extends AbstractLabelCycleReporter {
-  private static final Predicate<SkyKey> IS_SUPPORTED_SKY_KEY =
+class TargetCycleReporter extends AbstractLabelCycleReporter {
+
+  private static final Predicate<SkyKey> SUPPORTED_SKY_KEY =
       Predicates.or(
+          SkyFunctions.isSkyFunction(SkyFunctions.CONFIGURED_TARGET),
+          SkyFunctions.isSkyFunction(SkyFunctions.ASPECT),
+          SkyFunctions.isSkyFunction(SkyFunctions.LOAD_STARLARK_ASPECT),
           SkyFunctions.isSkyFunction(SkyFunctions.TRANSITIVE_TARGET),
           SkyFunctions.isSkyFunction(SkyFunctions.PREPARE_ANALYSIS_PHASE));
 
-  TransitiveTargetCycleReporter(PackageProvider packageProvider) {
+  TargetCycleReporter(PackageProvider packageProvider) {
     super(packageProvider);
   }
 
+  @Override
+  protected boolean shouldSkipOnPathToCycle(SkyKey key) {
+    return SkyFunctions.PREPARE_ANALYSIS_PHASE.equals(key.functionName());
+  }
+
   @Override
   protected boolean canReportCycle(SkyKey topLevelKey, CycleInfo cycleInfo) {
-    return Iterables.all(Iterables.concat(ImmutableList.of(topLevelKey),
-        cycleInfo.getPathToCycle(), cycleInfo.getCycle()),
-        IS_SUPPORTED_SKY_KEY);
+    return SUPPORTED_SKY_KEY.apply(topLevelKey)
+        && cycleInfo.getPathToCycle().stream().allMatch(SUPPORTED_SKY_KEY)
+        && cycleInfo.getCycle().stream().allMatch(SUPPORTED_SKY_KEY);
   }
 
   @Override
-  protected Label getLabel(SkyKey key) {
-    return ((TransitiveTargetKey) key).getLabel();
+  public String prettyPrint(SkyKey key) {
+    if (SkyFunctions.isSkyFunction(SkyFunctions.CONFIGURED_TARGET).apply(key)) {
```

Comment:
Why not check if instanceof? I think that would be clearer here, unless there's a good reason not to.

Reply:
good idea